rust/core-lib/src/watcher.rs

```diff
@@ -213,22 +213,35 @@ impl FileWatcher {
     }
 
     /// Takes ownership of this `Watcher`'s current event queue.
-    pub fn take_events(&mut self) -> VecDeque<(WatchToken, DebouncedEvent)> {
+    pub fn take_events(&mut self) -> VecDeque<(WatchToken, Event)> {
         let mut state = self.state.lock().unwrap();
         let WatcherState { ref mut events, .. } = *state;
         mem::replace(events, VecDeque::new())
     }
 }
 
 impl Watchee {
-    fn wants_event(&self, event: &DebouncedEvent) -> bool {
-        use self::DebouncedEvent::*;
-        match *event {
-            NoticeWrite(ref p) | NoticeRemove(ref p) | Create(ref p) | Write(ref p)
-            | Chmod(ref p) | Remove(ref p) => self.applies_to_path(p),
-            Rename(ref p1, ref p2) => self.applies_to_path(p1) || self.applies_to_path(p2),
-            Rescan => false,
-            Error(_, ref opt_p) => opt_p.as_ref().map(|p| self.applies_to_path(p)).unwrap_or(false),
+    fn wants_event(&self, event: &Event) -> bool {
+        match &event.kind {
```

Comment:
Because the event no longer stores the paths in the same enum as the type, we can dramatically simplify this function, to something like:

```rust
fn wants_event(&self, event: &Event) -> bool {
    event.paths.iter().map(|p| self.applies_to_path(p)).any()
}
```

Reply:
Yeah. But Im taking advantage of `wants_event` to do path's length check. And the path's length will change based on event kind.

Reply:
I think my code should work regardless of the length of `paths`.

Reply:
To make this more clear: my code checks every path in `paths`, and returns `true` if `self.applies_to_path` is true for *any* of them. Do we need any more logic than this?

Reply:
Sorry I maybe making this harder than it needs to be but I can think of two scenarios we need to avoid:
1. Imagine we get a rename event with ONLY from path. to is missing. Probably a bug in notify. 
This `event.paths.iter().map(|p| self.applies_to_path(p)).any()` might/ might not return true. If it does, we are guaranteed to panic down the line. If at this point, `wants_event` returns false...we can avoid panic. 
2. For other events (create, modify, remove) what if paths len is not 1? Other suggestion `event.paths.first().map(|p| self.applies_to_path(p)).unwrap_or(false)` will work even if path > 1. But we need to return false here too. 

So to avoid index out of bound panics, we should check
1. For create, modify and remove events we check it they have exactly one path. 
2. For rename events, check if it has exactly two. 

If there is a more rustic way to do this....I would be happy to look in to it.

Reply:
Okay, I see some of what you're saying. That said, I think the larger problem is that we are unsure of the API guarantees of the `notify` crate. Is it *possible* for a `Rename` event to arrive with less than 2 paths? If it is, then it is totally reasonable for other code of ours to want to receive that event and handling it; if other code is assuming there are always 2 paths, I think that's a mistake.

I think it might be okay if we ignore events with 0 paths, but it might be worth logging if this is happening.

So basically: I think we need to correctly handle a variable number of paths, if that is something that the API provides under normal conditions.

Reply:
> Is it possible for a Rename event to arrive with less than 2 paths?

To be specific we use `RenameMode::Both` and I dont think so. See [this](https://docs.rs/notify/5.0.0-pre.1/notify/event/enum.RenameMode.html#variant.Both). For all other events notify doesnt say they will have exactly 1 path. So we do that check. 

> If it is, then it is totally reasonable for other code of ours to want to receive that event and handling it; 

This means putting conditions all over. We can avoid this. 

> if other code is assuming there are always 2 paths, I think that's a mistake.

Other code can safely assume there are 2 paths because `wants_event` gave us the guarantee. If you agree to this..then I dont think we need more changes to `wants_event`. Changes might be requried to make it more rustic. Correct me if Im wrong. 

> I think it might be okay if we ignore events with 0 paths, but it might be worth logging if this is happening.

Agree. Making this change.